dvjsharma/Drawn2Shoe
# feat: Implemented User Profile Page issue: #143

## Description

This PR introduces several new features and components to the User Profile page, enhancing the user experience and functionality. The following key features have been implemented:

### Changes


**1. User Profile Page Layout**

   -   Created a structured layout for the User Profile page.
   -   Implemented navigation to different sections such as Account Details, Wishlist, Purchased Items, Payment Methods, and 
       Address.
![image](https://github.com/user-attachments/assets/a4fba3f3-d80f-4fc2-ae0e-6d936ec2a67c)

**2. Account Details Component**

   -   Displays user information including name, email, and profile picture.
   -   Allows users to update their profile details and upload a new profile picture.
![image](https://github.com/user-attachments/assets/32402b32-0945-4e5d-aa8f-ccae4954c8bc)

**3. Wishlist Component**

   -  Displays the user's wishlist items in a visually appealing card layout.
   - Uses DaisyUI and Tailwind CSS for styling.
   - Includes a heart button on product cards to add/remove items from the wishlist.
   - Fetches and displays the wishlist items dynamically.
   - Implements pagination for the wishlist, showing 5 items per page in a three-column layout.
![image](https://github.com/user-attachments/assets/39ea5e5f-0021-4a71-b24e-31177919a9b0)
![image](https://github.com/user-attachments/assets/adebf5d8-148a-4899-b2e8-03a565796b2f)

**4. Purchased Items Component**

   - As of now I have added a empty image in this component.    
![image](https://github.com/user-attachments/assets/7c85a9c0-6caf-4bdf-9e05-a0a11faa45cb)

**5. Payment Methods Component**

   - As of now this also implemented as empty.   

![image](https://github.com/user-attachments/assets/9dcb67b2-e02c-4962-99ab-d8f2147b3dce)

**6. Address Component**

   - Displays user's saved addresses.
   - Ensures proper validation of address fields.
![image](https://github.com/user-attachments/assets/9f3d5df1-cd7c-4dc5-9c9f-a186cf91a4d3)

**7. Prisma Schema Updates**

   - Created a new schema for handling wishlist items.
   - The Wishlist model includes fields for id, userId, and productId

           

> 
>             model Wishlist {
>                 id               Int      @id @default(autoincrement())
>                userId         String
>                productId   Int
>                product      Product  @relation(fields: [productId], references: [productId])
>               user             Mainuser @relation(fields: [userId], references: [email])
>                @@unique([userId, productId])
>              }
>           

**8. Wishlist Controller**

   - `addToWishlist` Function
       - Adds a product to the wishlist.

       - Removes the product from the wishlist if it already exists.

  - `getWishlist `Function

     - Fetches the user's wishlist items.
  
**9. PrivateRoute Component**

   - `PrivateRoute`  Component

       - Created a `PrivateRoute` component to protect routes that should only be accessible to logged-in users.
      - Redirects users to the login page if they are not authenticated.


##Bug Fixes

1. Addressed issues related to adding/removing items from the wishlist.
2. Improved state management to ensure the correct display of wishlist items.
3. Added a border to navbar.
4. Added a profile image on navbar at place of user name after login.
5. Added a dropdown on the profile avatar of `user profile page` and `logout`.
6. Implemented a `privateroute` component for preventing user to visit `profile page` , `cart page` without login.

Before:

After:
![image](https://github.com/user-attachments/assets/7da8c97a-5012-4666-bc56-5dc5dd717046)
![image](https://github.com/user-attachments/assets/aec6afe3-edb3-4581-b8e6-8957baa92cd9)


## How to test

**1. Setup:**

 - Ensure that the server and client are running.
 - Ensure you are logged in to access the User Profile page.

**2. Accessing User Profile Page:**

 -  Click on the profile picture in the navbar.
 -  Verify that the User Profile page loads with the appropriate sections (Account, Wishlist, Purchased Items, etc.).

**3. Testing Wishlist Functionality:**

 -  Navigate to the product listing page.
 -  Click the heart button on a product card to add it to the wishlist.
 -  Verify that the heart button turns red.
 -  Reload the page and verify that the heart button remains red for items in the wishlist.
 -  Navigate to the Wishlist section in the User Profile page.
 -  Verify that the added product appears in the wishlist.
 -  Remove a product from the wishlist by clicking the heart button again.
 -  Verify that the product is removed from the wishlist and the heart button turns gray.

**4. Testing PrivateRoute:**

 -  Try to access the User Profile page while logged out.
 -  Verify that you are redirected to the login page.
 -  Log in and try to access the User Profile page again.
 -  Verify that you can now access the User Profile page.

It fixes the issue #143 .
